### DISCUSSION TITLE
Fixing that pesky building type quest

### DISCUSSION THREAD

**Discussion by smichel17** [streetcomplete/StreetComplete#3558] [Feedback]
Building type is a problematic quest, even for experienced mappers. I've answered it 2313 times (my most-answered quest), and I still have extreme trouble deciding between `detached` and `apartments` sometimes, even after the long discussion on this topic in https://github.com/streetcomplete/StreetComplete/issues/2257. @RubenKelevra is another experienced mapper with the [same problem](https://github.com/streetcomplete/StreetComplete/issues/3248#issuecomment-974727125).

I can provide a bunch of pictures I took recently of places where residential is just impossible to decide. I've left them out for now because I think the issue is really with the entire quest, not just residential buildings. `shop` vs `office` is also frequently impossible to figure out without some knowledge of the building's history. And that's exactly the problem: the line between **building type vs building use**.

The latter is often much easier to decide. I can tell you whether a building is being used as a shop or office. I can tell you that a building is residental. I can even tell you how many different units a building has. I cannot for the life of me tell you whether the building was *constructed* as an apartment building vs a detached house where someone later added an accessory apartment. I cannot tell you whether it's a semi-detached house with a shared wall/roof or something crazy like the [house I used to live in](https://github.com/streetcomplete/StreetComplete/issues/2685#issuecomment-809473943). At least, not without going inside. It seems *incredibly likely* that **inexperienced mappers** would **mix up type vs use,** especially since SC does not have a building:use quest.

**I'll post some ideas for solutions as replies to this thread**, since I think there are **multiple ways this could be solved.** But I just want to capture the idea that **this quest needs major changes**. As it is, I'm worried about the harm it does by creating [garbage data](https://www.openstreetmap.org/changeset/114235101), especially (but not only) for residential buildings. I don't propose disabling it, but only because it enables the address quest, and it's been around for long enough that an extra month or two isn't going to make that much of a difference in the impact.

**Comment by RubenKelevra** [comment]
I like I add that the quest does not fulfil the

- [ ] 🐿️ Easily answerable by everyone from the outside but a survey is necessary

requirement.

Example:

`building:use` like there's a shop and two residents living there is easily observable.

But that the building was 80 years ago constructed as a train station isn't even remotely possible to observer anymore.

**Reply by matkoniecz** [reply]
Note that it is not asking about original purpose, or current style, it is asking about current building tyle.

**Reply by smichel17** [reply]
`tyle` seems like a typo— `type` or `style`?

**Reply by mnalis** [reply]
@smichel17 maybe `tyle` is combination of both, as both make sense. I'd prefer `stype` though :smile: 

@matkoniecz while that indeed seems to be what wiki says those days, it is **_highly unintuitive_** (even to seasoned OSM mappers, much less target SC audience! and few words in the quest are not going to fix that IMHO) and also somewhat controversial. 

I certainly know I mapped my first several thousand buildings according to their current purpose (instead of their original architectural style) before I even became aware of the `building` / `building:use` schism. And I know quite a few mappers who are not aware or think/map otherwise. 

And even after I become enlightened, the SC quest is often complicated, so I often prefer to use widest possible answer (eg. `building=residential` or `building=commercial`)  just so I can get to other much more important quests like `address` quest. My only regret is that answering `residential` does not get cached for some reason and I have to confirm such less precise answer, so I have to go through more clicks to select it -- and even then, it is still very often (to me) **_preferable to mental load_** of trying to figure if that is really `building=shed` or is it actually `building=garage` which has been converted to `building:use=shed` (or commercial/apartments, or house/detached, or hangar/marketplace etc... - I really just want to map an address of that building, I'm not interested in its history in vast majority of the cases)

**Reply by westnordost** [reply]
This. The app does **deliberately** not ask about hte original purpose. The worst thing that can happen that a building "wrongly" gets an e.g. `building=apartments` even though it **ought** to be e.g. `building=industrial + building:use=apartments` according to some bean counters in OSM. Well, then `building=apartments` is still an improvement in precision over `building=yes`.

**Reply by RubenKelevra** [reply]
In this case, I would consider this as a bug. The wiki states that the building key is the original type of building.

"The most basic use is building=yes, but the value may be used to classify the type of building. Note that it may be not the same as the building's current use (tagged using building:use=*). For example, a hospital building that is abandoned or repurposed to be a marketplace is still a building=hospital, and to mark active hospitals amenity=hospital is used."

There's no reason to tag a `bulding=yes` as `building=apartments` if it's say an old factory. If the app asks for the current use, it should use `building:use=*`

**Comment by matkoniecz** [comment]
> constructed as an apartment building vs a detached house where someone later added an accessory apartment

I would consider reducing choices to `residential` with `house` `aparments` subtypes. I am not convinced that remaining types are actually worth distinguishing.

**Reply by mnalis** [reply]
Maybe clicking on the category name (like `residential`, `commercial`, `civic`) should immideately map the building as that generic type, and if someone wants to add more details, then they should jump through more hoops (eg. click on triangle on `residential` to expand a sub-menu where they can click on more detailed type). Because, some people might dislike if you take away choices that were once before given to them.

Or (IMHO, preferable) make a building type quest optional (maybe even disabled by default?) so people who don't care about building types / heights / roof types etc. are not blocked from solving quest they deem much more important (like address quests - https://github.com/streetcomplete/StreetComplete/issues/2464) ?

**Reply by matkoniecz** [reply]
> Or (IMHO, preferable) make a building type quest optional (maybe even disabled by default?) so people who don't care about building types / heights / roof types etc. are not blocked from solving quest they deem much more important (like address quests - #2464) ?

Asking address quests on `building=yes` would be

- confusing
- requiring people to specify building type anyway
- adding `noaddress=yes` to garages and so on seems weird, unusual and confusing to me (and I am dubious is it something that would be accepted by wider community)
- it would also dilute address quest - nowadays when I see it then I know that it is likely to be quite important why building quest can be treated as low priority

**Reply by mnalis** [reply]
@matkoniecz the first three issues were rehashed already before (so, I won't duplicate that discussions); but I'm curious if you can expand what you meant by 4th bullet, as I didn't quite manage to parse that?

**Reply by matkoniecz** [reply]
Current situation:

every address quest indicates that address is missing, so solving it will add address (sometimes OSM data is simply wrong what is also an important fix)

After asking address quest on `building=yes`:

vast majority will be asked on `building=yes` that are garages, sheds etc. Only tiny minority of quests will result in adding address or fixing broken OSM data

**Reply by andrewharvey** [reply]
I find the different residential dwelling types are very useful, they paint a good picture of what the street/suburb/city composition of dwelling types are, are they detached houses, duplex, terraces, apartment blocks. And eventually once we get the wiki tagging sorted be able to see granny flats.

**Comment by smichel17** [comment]
What are the main uses for the building type quest? The [original issue](https://github.com/streetcomplete/StreetComplete/issues/25) is before the quest template was added.

As far as I can tell, it's basically not useful *on its own*, except maybe for some things like zoning or population density analysis; the main use is identifying buildings which could use additional detail, like whether a building is a candidate for the address quest or whether there should be a shop tagged in this building.

**Reply by mnalis** [reply]
Main use seems to be that SC does not want to ask for `house address` quests on ineligible buildings (like sheds and garages). 

See discussion in https://github.com/streetcomplete/StreetComplete/issues/2464 which also proposes other ways to deal with that issue and thus remove requirement that less useful `building type` must be solved before much more important quests like `house address` are unlocked.

That issue was tentatively closed, with note that might be reopened later https://github.com/streetcomplete/StreetComplete/issues/2464#issuecomment-783351852

**Comment by smichel17** [comment]
I got one idea from the [`Key:building:use`](https://wiki.openstreetmap.org/wiki/Key:building:use) wiki page, which says

> `building=*` can imply the building's function, so this tag may be unnecessary in normal cases:
> 
> - `building=apartments`
> - `building:use=residential` *– usually not worth adding: apartments implies residential*

So, what if we did this as a two-stage quest (two separate quests):

1. Tag `building:use` with broad categories, like `building:use=residential`.
2. Tag a more specific building type, like `building=apartments`, and remove the now-redundant `building:use` tag.

This would get a bunch of nice things:

- Greatly simplified building type quest, since only one category at a time will need to be displayed, based on `building:use`. Especially: Removes the need for the confusing "category view, where you can also select the category as an answer" UI.
    - As a result, we may be able to support a greater variety of building types & uses which have been requested in the past, without clogging up the UI.
- Building type quest can be moved down the priority list (maybe even disabled by default?) or selectively enables/disabled for certain answers of `building:use`, because it is no longer a prerequisite for the address quest.
- Right now, I am reluctant to answer with the generic `building=residential`, because I don't want to lose the chance to tag something more specific. This avoids the problem, since now we can tag `building=residential` in the follow-up quest (and remove `building:use=residential`) if that is actually the most specific value possible.

The main — and significant — downside is that I'm not sure `building:use` is well-defined enough to cover all of our use cases (pun intended), so it may require a tagging proposal to define additional tags. If necessary, I *might* be willing to break my normal abstinence from upstream matters and do this :grimacing:.

**Reply by mnalis** [reply]
While I agree that `building:use` is more important than `building` (which is also likely as important as for example its period architecture - eg. is it `gothic` / `barouqe` / `secession` / `socialist`...), I really don't care _much_ about either. I grudgingly solve that `building type` quest (eh, to be politically correct, let's say "of very low practical usefulness"), in vast majority of cases only because it is prerequisite for solving super-important `house address` quest. 

If that changes so I'd have to solve **two** "useless" quests in order to get to the important one, I would not consider that an improvement :sob:. 

If it however only replaces dependency on `building` with dependency on `building:use`, than it is OK, and basically (for UX, not for actual tags used) somewhat close to [@matkoniecz idea](https://github.com/streetcomplete/StreetComplete/discussions/3558#discussioncomment-1744074) of simplifying what answers to offer to the users.  (While that would not solve the problem of less import quest being a prerequisite of more important one, at least if would not make a situation significantly worse)

Also note that big part of current problems with `building=*` tag remains with `building:use` - specifically, that the buildings are often multi-use: apartments+shops etc; where the "solution" seems to be "just ignore the minorities", with the (even more horrible) alternative of completely overlapping `building:part`s.

If might also further confuse the users: asking about building use first, and its building style later, with some of the answers being same (like `residential` etc).

**Reply by smichel17** [reply]
> If it however only replaces dependency on `building` with dependency on `building:use`

Yes, this is what I meant.

> somewhat close to [@matkoniecz idea](https://github.com/streetcomplete/StreetComplete/discussions/3558#discussioncomment-1744074) of simplifying what answers to offer to the users

Yes. The main difference is that this idea **does not prevent adding a more specific type later.**

If we only simplify the building type quest, and tag `building=residential` then we cannot add another quest saying "What specific type of residential building is this?" because there's no equivalent tagging for "There is no more specific type."— if we left `building=residential` in place, the quest would immediately be asked again.

If we tag `building:use=residential` and there is no more specific type, then we can tag `building=residential` (and remove the now-redundant `building:use=residential`) to solve the quest.

**Comment by smichel17** [comment]
> My only regret is that answering residential does not get cached for some reason

@mnalis I noticed this when I was implementing https://github.com/streetcomplete/StreetComplete/pull/3373. I suspected it was intentional, so I preserved the behavior, but it should be easy for me to change, if @westnordost agrees it's a good idea.

**Reply by mnalis** [reply]
@smichel17 see https://github.com/streetcomplete/StreetComplete/issues/3600

**Comment by RubenKelevra** [comment]
I would reduce the quest to a row of pictures (like the wiki does) with the most common building types found in the country.

This would lead to a `building:use=*` tag instead of a `building=*` tag in most cases.

While `building=*` makes sense for say a garage it doesn't make sense for a small house that might have 2-3 families or just one living there.

It's too complex to go to each door and have a look at how many doorbell buttons there are.

Having a `building:use` tag would allow us to create shop nodes inside of buildings automatically if the user answers that there are shops but no shop node exists inside the building.

**Comment by smichel17** [comment]
Can someone find or compile a comprehensive list of `building:` subkeys? Searching for `key:building:` on the wiki yields a bunch, but also a bunch of noise. So far I have these, in *rough* order of relevance to this quest.

- [`building`](https://wiki.openstreetmap.org/wiki/Key:building)
- [`building:use`](https://wiki.openstreetmap.org/wiki/Key:building:use)
- [`building:architecture`](https://wiki.openstreetmap.org/wiki/Key:building:architecture)
- [`building:flats`](https://wiki.openstreetmap.org/wiki/Key:building:flats)
- [`building:levels`](https://wiki.openstreetmap.org/wiki/Key:building:levels)
- [`building:min_level`](https://wiki.openstreetmap.org/wiki/Key:building:min_level)
- [`building:condition`](https://wiki.openstreetmap.org/wiki/Key:building:condition)
- [`building:material`](https://wiki.openstreetmap.org/wiki/Key:building:material)

**Reply by mnalis** [reply]
Taginfo says those are most popular `building:*` keys: https://taginfo.openstreetmap.org/search?q=building:#keys

**Comment by smichel17** [comment]
From https://github.com/streetcomplete/StreetComplete/issues/3793#issuecomment-1047894894, I just learned that

> In [the last year](https://taghistory.raifer.tech/#***/house/), the `house=*` has gained some traction (mostly in Ireland):
> 
> https://wiki.openstreetmap.org/wiki/Key:house

That seems like it could solve some issues. Not everything, because I still need to decide between house & apartments, but being able to delegate "what type of house is this?" to a second level quest would make the building type *much* simpler.

If this tag gains enough traction that SC can implement it, I would be highly in favor of eliminating all of the house subtypes which can be expressed with `house=*`.

**Comment by smichel17** [comment]
In a previous discussion about building types (I don't remember which), somebody suggested that you can tell the number of units in the house by checking the number of electric/gas meters. Since then, I've found this to be a pretty good indicator of whether a building is a single-family detached house or an apartment. Actually, it's one of the *only* reliable indicators I've found (the other one: does it have two separate mailboxes or numbers next to the door?). Since I have been using this heuristic, I no longer have trouble answering this question. So, I think we can improve this by changing the image or text for "apartment building". I'll open a new issue when I have a concrete suggestion.

**Reply by westnordost** [reply]
No, don't, it will be rejected. This only works if the gas/electric meters are always outside the building, which they are not.

**Reply by smichel17** [reply]
On one hand, you are right. On the other, if there **are** multiple meters outside, then it is surely an apartment, so it is useful to know about. I didn't think about it before someone mentioned it here.

Determining whether something is a 2-unit apartment or a single-family home is always like this. You look for one of the things that indicates it's an apartment; if none of them are present, then you just have to guess it must be a detached house. At least, that's my experience.

**Reply by mnalis** [reply]
I guess availability of such information is country-dependent. For example in Croatia, in vast majority of the apartment cases, you won't be able to see _neither_ multiple electric meters _nor_ multiple gas meters _nor_ multiple postboxes from outside - they'd all be inside a building itself. And especially, electric and especially gas meters would be _inside individual apartments_ themselves (i.e. unless the meters are new wireless-types, the electric/gas company employee has to ring up each individual resident in order to be allowed to read their meters.)

So while I'm happy that @smichel17 has found a working solution for himself, it doesn't work for me at all, so I'd still very much like an option to at least _more easily_ mark building as generic `building=residential` (https://github.com/streetcomplete/StreetComplete/issues/3600) -- or (of course, even better) skip the pesky `building type` quest entirely (https://github.com/streetcomplete/StreetComplete/issues/2464)

**Reply by smichel17** [reply]
@mnalis To clarify:

1. I agree about the generic version, but would rather move the residential/house sub-categories into their own quest. See above https://github.com/streetcomplete/StreetComplete/discussions/3558#discussioncomment-2236481.

2. Note: I was just talking about **2-family** apartments. There's also larger apartment buildings, which typically have a single entrance, but they are easier to identify by other means (like having a lobby, or a buzzer/intercom outside to ask someone in the apartment to unlock the entrance to the shared stairwell.

3. My area definitely has some buildings like you describe. The problem is, I cannot **reliably classify** them. So, I have been sticking to my heuristics (if I do not see one of the indications of an apartment, then I tag as detached-house) because it **makes my tagging consistent**. Before, my tags were so inconsistent as to be meaningless, since each unclear house was tagged basically with my best guess.

    - At least in the current way, if I am wrong, I will be consistently wrong in the same way, so it's possible to fix the data in a smarter way than just "remove all building types added by @smichel17 between these dates". For example, now I am confident that all of my `building=apartment` answers are correct, while `building=detached` may still be wrong.

**Reply by mnalis** [reply]
@smichel17 TL;DR - I don't think it would work, because :

- I **like the idea** of _drastically reducing_ `building type` choices, and asking details in separate quest. But, in my opinion, asking **only** for `building=house` subtypes to mark in `house=*` is **too little to help** - as it just doubles the work for people actually caring about details of building type (who now have to answer two quests), without making (noticeably) less work for those who do not really care about it (and just want to get to important `address quest`, for example). So, not much improvement for one class of the users, and would probably make things worse for other class of users.

- Now, if `building type quest` offered **only 6 choices** which all fit on single screen: `residential building` / `industrial building` / `retail` / `shed` / `garage` / `something else` (where `something else` would ask all the mega-detailed choices we have today) that would be more usable to me, as I'd be able to solve that one instantly in 98% of the cases in order to get `address quest` (which is the one I really care about), and just skip the remaining 2%.  (And yes, I am aware the current "recents" category would look mostly like this after little usage -- if not for the fact that it does not currently cache generic types `residential` / `industrial` / etc. due to #3600)  

- But if it would just shorten current megalist by mere 5-10% and then add another quest to fill more details for those few `house` subtypes, then (at best) it doesn't really do anything for me.

- I don't really like wasting time on building type quest, especially as I don't see much use for it (especially as we have [landuse=*](https://wiki.openstreetmap.org/wiki/Key:landuse) which covers most use cases IMHO) ... So, if it takes more than 1.5 seconds to precisely and surely determine, in my case it's going to be generic `building=residential` (instead of me breaking my head over whether it's multi-resident apartment, two-resident apartment, semi-detached, detached, house in a row of houses, regular house or bungalow or farmhouse or whatever  -- **it's just a place where people live**, for [bikeshedding](https://en.wiktionary.org/wiki/bikeshedding) sake!)

- at some time there was idea that each quest would have associated description for newcomers **why specific quest is it useful** (i.e. what data consumers use it for what purpose). Unfortunately I can't find that issue right now to link it, but I think it would be great idea, and not just for newcomers -- I'd personally _love_ to read that description explaining why we (or anybody) care if building is marked as `semi-detached` or as `detached` or as `house` or as `bungalow`. We are overloading our users with [too many choices to choose from](https://www.fastcompany.com/3031364/why-having-too-many-choices-is-making-you-unhappy), and that is a **real problem even if such choices were to make a big difference** -- and just wrong when the differences between choices are practically useless.

**Reply by smichel17** [reply]
> So, if it takes more than 1.5 seconds to precisely and surely determine, in my case it's going to be generic `building=residential` (instead of me breaking my head over whether it's multi-resident apartment, two-resident apartment, semi-detached, detached, house in a row of houses, regular house or bungalow or farmhouse or whatever  -- **it's just a place where people live**, for [bikeshedding](https://en.wiktionary.org/wiki/bikeshedding) sake!)

I agree, but **in my opinion, the subtypes of houses *are* actually what makes it take more than 1.5s to answer.**

I cannot find a clear, meaningful difference between a detached, single-family house vs a detached, multi-family house like the ones pictured in https://github.com/streetcomplete/StreetComplete/issues/2257#issuecomment-728676151=. However, I definitely can vs a Manhattan-style apartment building with an elevator, probably a lobby, and maybe even a security guard at the front desk.

- Clear: The city-style apartments generally have cookie-cutter layouts, with a single, linear dividing wall between apartments. As opposed to apartments with a layout like https://github.com/streetcomplete/StreetComplete/issues/2685#issuecomment-809473943.

- Meaningful: It boils down to whether it's exclusively **private space**. Even if a detached, multi-family "apartment building" had a shared lobby, I would not feel comfortable letting myself in to warm up while out on a walk on a cold day. Stopping in a city apartment's lobby? No problem (if it's allowed).

Something like the [Garden Apartments](https://www.openstreetmap.org/way/248564943#map=19/40.74833/-73.75256) stretches this classification a little — they are almost like row houses, each near-identical 2-story house with its own entrance, but I would definitely tag them apartments — but if you think of the little courtyards they are arranged around as the equivalent of the public lobby, then it fits well, like a Manhattan-style apartment building laid flat on its side.

My suggestion would take the Residential types down to something like `house` (even if split into a few rental units), `apartment building`, `houseboat`, `mobile home`, and `dormitory`. I think that would be easy to answer, useful, and detailed-enough that I wouldn't care to spend my time answering the "house type" quest.
